quinn-proto/src/packet.rs

```diff
@@ -219,6 +382,65 @@ impl Header {
     }
 }
 
+pub struct PartialEncode<'a> {
+    header: &'a Header,
+    pn: Option<(usize, usize)>,
+}
+
+impl<'a> PartialEncode<'a> {
+    pub fn finish(self, buf: &mut [u8], pn_key: &PacketNumberKey, header_len: usize) {
+        let PartialEncode { header, pn } = self;
+        let payload_len = (buf.len() - header_len) as u64;
+        let (sample_offset, pn_pos, pn_len) = match header {
+            Header::Short { dst_cid, .. } => {
+                let mut sample_offset = 1 + dst_cid.len() + 4;
+                let packet_length = buf.len();
+                if sample_offset + pn_key.sample_size() > packet_length {
+                    sample_offset = packet_length - pn_key.sample_size();
+                }
+                let (pn_pos, pn_len) = pn.unwrap();
+                (sample_offset, pn_pos, pn_len)
+            }
+            Header::Initial {
+                dst_cid,
+                src_cid,
+                token,
+                ..
+            } => {
+                let sample_offset = 10
+                    + dst_cid.len()
+                    + src_cid.len()
+                    + varint::size(payload_len).unwrap()
+                    + varint::size(token.len() as u64).unwrap()
+                    + token.len();
+                let (pn_pos, pn_len) = pn.unwrap();
+                (sample_offset, pn_pos, pn_len)
+            }
+            Header::Long {
+                dst_cid, src_cid, ..
+            } => {
+                let sample_offset =
+                    10 + dst_cid.len() + src_cid.len() + varint::size(payload_len).unwrap();
+                let (pn_pos, pn_len) = pn.unwrap();
+                (sample_offset, pn_pos, pn_len)
```

Comment:
Probable missing bounds check

Reply:
I think this is covered by the `debug_assert!()`s, below, unless you think they should be non-debug asserts? This is one of those cases where things can only go wrong if we've got bugs in our own code, so I didn't think we'd need to do much more.

Reply:
Have another look, I think it'll make more sense now.